File: packages/engine/src/engine/roles/doctor.py
import engine.events as events
from engine.events import ACTION_EVENTS
from engine.models import Player

# from engine.roles import Actor
from engine.roles.actor import Actor, Town
from engine.utils.logging import logger


class Doctor(Town):
    tags = ["any_random", "town_random", "town_protective"]

    def __init__(self, player: Player, settings: dict = dict()):
        super().__init__(player)

    def find_possible_targets(self, actors):
        num_targets = 1
        self.possible_targets = []
        for i in range(num_targets):
            self.possible_targets.insert(
                i,
                [actor for actor in actors if actor.alive and actor != self],
            )

    def action(self) -> None:
        target = self.targets[0]
        logger.info(f"{self} will attempt to heal {target}")
        self.visit(target)
        target.doctors.append(
            self
        )  # Add self into the list of doctors protecting this target

    def revive_target(self, target: Actor) -> None:
        logger.info(f"{self} revives {target}")

        # Event group for the revival
        revive_event_group = events.GameEventGroup(group_id="doctor_revive")

        # Inform the doctor that the target was revived
        revive_event_group.new_event(
            events.GameEvent(
                event_id="doctor_revive_success",
                targets=self.player.id,
                message="Your target was attacked last night, but you successfully revived them",
            )
        )

        # TODO: Possibly inform all other doctos that the target was revived

        # Inform the player that they were revivied
        revive_event_group.new_event(
            events.GameEvent(
                event_id="revive_by_doctor",
                targets=[target.player.id],
                message="You were revived by a doctor. Rock on",
            )
        )

        ACTION_EVENTS.new_event_group(revive_event_group)

File: packages/engine/src/roles/doctor.ts
import { z } from 'zod';
import { EventIds } from '../constants';
import { GameEvent, GameEventGroup } from '../events';
import { Town, type Actor, type ActorContext, type ActorState } from './actor';
import { RoleTags } from './role';

export const DoctorSettingsSchema = z.object({}).strict();

export type DoctorSettings = z.infer<typeof DoctorSettingsSchema>;
export type DoctorSettingsInput = z.input<typeof DoctorSettingsSchema>;

export class Doctor extends Town {
	static override tags = [
		...super.tags,
		RoleTags.TownProtective,
	] as const;

	static override roleName = 'Doctor' as const;
	static override roleKey = 'doctor' as const;

	static override priority = 1;
	static settingsSchema = DoctorSettingsSchema;
	static description = 'Town protective role that can heal one target each night.';

	constructor(
		input: ActorState,
		settings: DoctorSettingsInput = {},
		context: ActorContext,
	) {
		super(input, context);
		DoctorSettingsSchema.parse(settings);
	}

	override findPossibleTargets(actors: Actor[] = []) {
		return this.setSingleTarget(actors, (actor) => actor.alive && actor !== this);
	}

	override action() {
		const target = this.targets[0];
		if (!target) return;
		this.logger.info(`${this.toString()} will attempt to heal ${target.toString()}`);
		this.visit(target);
		target.doctors.push(this);
	}

	reviveTarget(target: Actor) {
		this.logger.info(`${this.toString()} revives ${target.toString()}`);
		const reviveEventGroup = new GameEventGroup(EventIds.DOCTOR_REVIVE);
		reviveEventGroup.newEvent(
			new GameEvent(
				EventIds.DOCTOR_REVIVE_SUCCESS,
				[this.input.id],
				'Your target was attacked last night, but you successfully revived them',
			),
		);
		reviveEventGroup.newEvent(
			new GameEvent(
				EventIds.REVIVE_BY_DOCTOR,
				[target.input.id],
				'You were revived by a doctor. Rock on',
			),
		);
		this.actionEvents.newEventGroup(reviveEventGroup);
	}
}
